modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/H2IndexCostedBase.java

```diff
@@ -16,57 +16,77 @@
 
 package org.apache.ignite.internal.processors.query.h2.opt;
 
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
 import java.util.ArrayList;
 import java.util.Arrays;
+
 import org.apache.ignite.IgniteLogger;
 import org.apache.ignite.IgniteSystemProperties;
+import org.apache.ignite.internal.processors.query.stat.ColumnStatistics;
+import org.apache.ignite.internal.processors.query.stat.ObjectStatisticsImpl;
 import org.apache.ignite.internal.util.typedef.F;
 import org.apache.ignite.internal.util.typedef.internal.LT;
 import org.gridgain.internal.h2.command.dml.AllColumnsForPlan;
 import org.gridgain.internal.h2.engine.Constants;
+import org.gridgain.internal.h2.engine.Session;
+import org.gridgain.internal.h2.expression.Expression;
+import org.gridgain.internal.h2.expression.condition.Comparison;
 import org.gridgain.internal.h2.index.BaseIndex;
 import org.gridgain.internal.h2.index.IndexCondition;
 import org.gridgain.internal.h2.index.IndexType;
 import org.gridgain.internal.h2.result.SortOrder;
 import org.gridgain.internal.h2.table.Column;
 import org.gridgain.internal.h2.table.IndexColumn;
 import org.gridgain.internal.h2.table.TableFilter;
+import org.gridgain.internal.h2.value.Value;
+import org.jetbrains.annotations.Nullable;
 
 /**
  * Index base.
  */
 public abstract class H2IndexCostedBase extends BaseIndex {
-    /** Const function. */
+    /**
+     * Const function.
+     */
     private final CostFunction constFunc;
 
-    /** Logger. */
+    private final GridH2Table tbl;
+
+    private final CostFunctionLast cfl = new CostFunctionLast();
```

Comment:
we don't need to store this because we could create this at the very moment of choosing a cost function